src/generate_pisa_index.cpp

```diff
@@ -21,19 +41,138 @@ std::ostream &write_sequence(std::ostream &os, gsl::span<T> sequence)
 }
 
 
+struct inverted_index {
+
+  std::unordered_map<uint32_t, std::vector<uint32_t>> m_documents {};
+  std::unordered_map<uint32_t, std::vector<uint32_t>> m_frequencies {};
+  std::map<uint32_t, uint32_t> m_document_sizes {};
+  std::vector<std::string> m_plain_terms {};
+ 
+  void add_document_lengths(const std::string doclen_file) {
+      std::ifstream input_lengths(doclen_file);
+      uint32_t doc_id;
+      uint32_t size;
+      while (input_lengths >> doc_id >> size) {
+          m_document_sizes[doc_id] = size;
+      }  
+      std::cerr << "Read " << m_document_sizes.size() << " document lengths."
+                << std::endl;
+  } 
+
+  void add_postings_list(const PostingsList& postings_list, uint32_t term_id) {
+
+      if (term_id % 10000 == 0) {
+          std::cerr << "Processing list " << term_id << "..." << std::endl;
+      }
+
+      std::string term = postings_list.term();
+      m_plain_terms.emplace_back(term);
+
+      uint32_t doc_freq = postings_list.df();
+      // uint32_t collection_freq = postings_list.cf();
+
+      std::vector<uint32_t> documents;
+      std::vector<uint32_t> frequencies;
+      m_documents[term_id].reserve(doc_freq);  
+      m_frequencies[term_id].reserve(doc_freq);
```

Comment:
I think it should be more efficient if you `resize` (or simply stuff the size in the constructor), and then use `i` below to index (it's already available anyway). Correct me if I'm wrong, but I think will still have some additional overhead for checking if it should resize, etc, while indexing is not checked.

Reply:
You are right here. I'll refactor this one.